### PR TITLE
Refactor gen_nondet_init

### PR DESCRIPTION
**Pull request by thk123**
As part of diffblue/test-gen#402 I split up `get_nodet_init` so this PR just does this to split up the changes (and so I can get this in sooner)

### CODE REVIEW

File: src/java_bytecode/java_object_factory.cpp [hunk 1]
```diff
@@ -411,6 +430,247 @@ void java_object_factoryt::gen_pointer_target_init(
...
+  const irep_idt &class_identifier,
+  bool create_dynamic_objects,
+  const pointer_typet &pointer_type,
+  const update_in_placet &update_in_place)
+{
+  const typet &subtype=ns.follow(pointer_type.subtype());
+  if(subtype.id()==ID_struct)
+  {
+    const struct_typet &struct_type=to_struct_type(subtype);
+    const irep_idt struct_tag=struct_type.get_tag();
```
Comment (mgudemann) [inline src/java_bytecode/java_object_factory.cpp]: could be reference

File: src/java_bytecode/java_object_factory.cpp [hunk 2]
```diff
@@ -411,6 +430,247 @@ void java_object_factoryt::gen_pointer_target_init(
...
+    // If this is a recursive type of some kind, set null.
+    if(!recursion_set.insert(struct_tag).second)
+    {
+      if(update_in_place==NO_UPDATE_IN_PLACE)
+      {
+        assignments.copy_to_operands(
+          get_null_assignment(expr, pointer_type));
+      }
+      // Otherwise leave it as it is.
```
Comment (mgudemann) [inline src/java_bytecode/java_object_factory.cpp]: from the comment it seems that an `else` case is missing

File: src/java_bytecode/java_object_factory.cpp [hunk 3]
```diff
@@ -411,6 +430,247 @@ void java_object_factoryt::gen_pointer_target_init(
...
+  // In particular the array field of a String should not be null.
+  bool not_null=
+    assume_non_null ||
+    ((class_identifier=="java.lang.String" ||
+      class_identifier=="java.lang.StringBuilder" ||
+      class_identifier=="java.lang.StringBuffer" ||
+      class_identifier=="java.lang.CharSequence") &&
+     subtype.id()==ID_array);
+
+    if(not_null)
```
Comment (mgudemann) [inline src/java_bytecode/java_object_factory.cpp]: indent wrong

File: src/java_bytecode/java_object_factory.cpp [hunk 4]
```diff
@@ -411,6 +430,247 @@ void java_object_factoryt::gen_pointer_target_init(
...
+  const exprt &expr,
+  bool is_sub,
+  irep_idt class_identifier,
+  bool skip_classid,
+  bool create_dynamic_objects,
+  const struct_typet &struct_type,
+  const update_in_placet &update_in_place)
+{
+  typedef struct_typet::componentst componentst;
+  const irep_idt struct_tag=struct_type.get_tag();
```
Comment (mgudemann) [inline src/java_bytecode/java_object_factory.cpp]: could be a reference

File: src/java_bytecode/java_object_factory.cpp [hunk 5]
```diff
@@ -411,6 +430,247 @@ void java_object_factoryt::gen_pointer_target_init(
 
 /*******************************************************************\
 
+Function: java_object_factoryt::gen_nondet_pointer_init
+
+  Inputs: `assignemnts` - the code block we are building with initilisation code
+          `expr`: lvalue expression to initialise
```
Comment (thk123) [inline src/java_bytecode/java_object_factory.cpp]: Missing `class_identifier`.

File: src/java_bytecode/java_object_factory.cpp (hunk 2 above)
Comment (thk123) [inline src/java_bytecode/java_object_factory.cpp]: Comment is from [original code](https://github.com/diffblue/cbmc/pull/990/files/9ca22853f62d418adeeab046bf8e1f5fcba77c02#diff-233653d9ec20537d004eef631225bfffL478). I don't _think_ there is supposed to be an else case. I read "leave as it is" to mean there is nothing to be done, can you think of a wording that doesn't imply to you an else clause?

**Comment by thk123** [comment]
@cesaro `get_nondet_init` as I understand, is responsible for generating the GOTO code that creates relevant NONDET instructions (e.g. the choices that CBMC makes) for java objects. For example, given some java:

```java
class A
{
  int a;
  B b;
}

class B
{
  int c;
}

class MainClass
{
  static void functionUnderTest(A a)
  {
  }
}
```

Then gen_nondet_init creates goto code that is kind of equivalent to

```C
B b = { .c = NONDET }
A a = { .a = NONDET, b = NONDET ? b : NULL }
```

From [Chris's talk](https://github.com/diffblue/doc-diffblue-common/wiki/Talks#chris-talk-about-cbmc-2017-06-06) it would be step 3.

**Comment by thk123** [comment]
@smowton done for #1027

File: src/java_bytecode/java_object_factory.cpp [hunk 6]
```diff
@@ -347,6 +366,226 @@ void java_object_factoryt::gen_pointer_target_init(
...
+///   instructions to
+/// \param expr: pointer-typed lvalue expression to initialise
+/// \param is_sub: If true, `expr` is a substructure of a larger object, which in
+///   Java necessarily means a base class. not match *expr (for example, expr
+///   might be void*)
+/// \param class_identifier: clsid to initialise @java.lang.Object. @class_identifier
+/// \param skip_classid: if true, skip initialising @class_identifier
+/// \param create_dynamic_objects: if true, use malloc to allocate objects; otherwise
+///   generate fresh static symbols.
+/// \param pointer_type - The type of the struct we are initalising
```
Comment (thk123) [inline src/java_bytecode/java_object_factory.cpp]: this should be `\param struct_type`

File: src/java_bytecode/java_object_factory.cpp [hunk 7]
```diff
@@ -270,7 +289,7 @@ exprt java_object_factoryt::allocate_object(
   }
 }
 
-/// Adds an instruction to `init_code` null-initialising `expr`.
+/// Adds an instruction to  init_code` null-initialising `expr`.
```
Comment (thk123) [inline src/java_bytecode/java_object_factory.cpp]: This was done in error

File: src/java_bytecode/java_object_factory.cpp [hunk 8]
```diff
@@ -347,6 +366,228 @@ void java_object_factoryt::gen_pointer_target_init(
...
+
+  // Similarly to above, maybe use a conditional if both the
+  // allocate-fresh and update-in-place cases are allowed:
+  if(update_in_place==NO_UPDATE_IN_PLACE)
+  {
+    assignments.append(new_object_assignments);
+  }
+  else
+  {
+    assert(update_in_place==MAY_UPDATE_IN_PLACE);
```
Comment (thk123) [inline src/java_bytecode/java_object_factory.cpp]: This assertion needs to become an invariant with a message as to why it is an invariant.

**Comment by thk123** [comment]
Applied comments made by Michael to this PR.

File: src/java_bytecode/java_object_factory.cpp (hunk 2 above)
Comment (peterschrammel) [inline src/java_bytecode/java_object_factory.cpp]: For me it's clear.

File: src/java_bytecode/java_object_factory.cpp [hunk 9]
```diff
@@ -347,6 +366,228 @@ void java_object_factoryt::gen_pointer_target_init(
...
+      code_assignt code(me, from_integer(0, me.type()));
+      code.add_source_location()=loc;
+      assignments.copy_to_operands(code);
+    }
+    else
+    {
+      INVARIANT(!name.empty(), "Each component of a struct must have a name");
+
+      bool _is_sub=name[0]=='@';
+#if 0
```
Comment (peterschrammel) [inline src/java_bytecode/java_object_factory.cpp]: These four lines can be removed.

**Comment by thk123** [comment]
@mgudemann Think I have done all your requests

File: src/java_bytecode/java_object_factory.cpp (hunk 2 above)
Comment (mgudemann) [inline src/java_bytecode/java_object_factory.cpp]: ok

**Comment by thk123** [comment]
Lint errors are unrelated, diffblue/test-gen#630 builds and has approval so merging.

**Pull request merged**